From 922e36bdd55c09a84a2c66013e09b77e4bcab1c0 Mon Sep 17 00:00:00 2001
From: 5656565566 <contact@home56.top>
Date: Fri, 28 Aug 2026 22:00:00 +0800
Subject: [PATCH] bionic: export clone-for-fork for NativeBridge guests

The NDK translation runtime resolves __clone_for_fork from the guest libc
during backend initialization. Keep the symbol hidden in regular libc builds,
but export it from the native-bridge guest variant.

Change-Id: I874b0b1724b95338cf70bbf49243a97be7840528
---
 libc/bionic/fork.cpp            | 2 +-
 libc/libc.map.txt               | 1 +
 libc/private/bionic_defs.h      | 3 +++
 3 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/libc/bionic/fork.cpp b/libc/bionic/fork.cpp
--- a/libc/bionic/fork.cpp
+++ b/libc/bionic/fork.cpp
@@ -34,7 +34,7 @@
 #include "private/bionic_fdtrack.h"
 #include "pthread_internal.h"
 
-__BIONIC_WEAK_FOR_NATIVE_BRIDGE_INLINE
+__BIONIC_WEAK_FOR_NATIVE_BRIDGE_EXPORT
 int __clone_for_fork() {
   pthread_internal_t* self = __get_thread();
 
diff --git a/libc/libc.map.txt b/libc/libc.map.txt
--- a/libc/libc.map.txt
+++ b/libc/libc.map.txt
@@ -1569,3 +1569,4 @@
+    __clone_for_fork; # arm arm64
     __accept4; # arm x86
     __adddf3; # arm
     __addsf3; # arm
diff --git a/libc/private/bionic_defs.h b/libc/private/bionic_defs.h
--- a/libc/private/bionic_defs.h
+++ b/libc/private/bionic_defs.h
@@ -38,10 +38,13 @@
 #define __BIONIC_WEAK_VARIABLE_FOR_NATIVE_BRIDGE __attribute__((__weak__))
 #define __BIONIC_WEAK_FOR_NATIVE_BRIDGE_INLINE \
   __BIONIC_WEAK_FOR_NATIVE_BRIDGE extern "C" __LIBC_HIDDEN__
+#define __BIONIC_WEAK_FOR_NATIVE_BRIDGE_EXPORT \
+  extern "C" __BIONIC_WEAK_FOR_NATIVE_BRIDGE __attribute__((__visibility__("default")))
 #else
 #define __BIONIC_WEAK_FOR_NATIVE_BRIDGE
 #define __BIONIC_WEAK_VARIABLE_FOR_NATIVE_BRIDGE
 #define __BIONIC_WEAK_FOR_NATIVE_BRIDGE_INLINE static inline
+#define __BIONIC_WEAK_FOR_NATIVE_BRIDGE_EXPORT static inline
 #endif
 
 #endif /* __BIONIC_PRIVATE_BIONIC_DEFS_H_ */
-- 
2.47.3
